flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/source/StreamingMonitorFunction.java

```diff
@@ -124,11 +126,33 @@ public void initializeState(FunctionInitializationContext context) throws Except
     if (context.isRestored()) {
       LOG.info("Restoring state for the {}.", getClass().getSimpleName());
       lastSnapshotId = lastSnapshotIdState.get().iterator().next();
+    } else if (scanContext.startTag() != null) {
+      Snapshot currentSnapshot =
+          scanContext.branch() == null
+              ? table.currentSnapshot()
+              : table.snapshot(scanContext.branch());
+      Preconditions.checkNotNull(
+          currentSnapshot,
+          "Don't have any available snapshot for branch " + scanContext.branch() + " in table.");
+
+      long currentSnapshotId = currentSnapshot.snapshotId();
+      long startSnapshotId = table.snapshot(scanContext.startTag()).snapshotId();
+      Preconditions.checkState(
+          SnapshotUtil.isAncestorOf(table, currentSnapshotId, startSnapshotId),
+          "The option start-snapshot-id %s is not an ancestor of the current snapshot.",
+          startSnapshotId);
+
+      lastSnapshotId = startSnapshotId;
     } else if (scanContext.startSnapshotId() != null) {
+      Snapshot currentSnapshot =
+          scanContext.branch() == null
+              ? table.currentSnapshot()
+              : table.snapshot(scanContext.branch());
       Preconditions.checkNotNull(
-          table.currentSnapshot(), "Don't have any available snapshot in table.");
+          currentSnapshot,
+          "Don't have any available snapshot for branch " + scanContext.branch() + " in table.");
```

Comment:
Also sorry if I missed it, but at some point shouldn't there be validation that the start tag is an ancestor of the end tag?

Reply:
They are eventually converted to snapshot id and checked during the `planTasks` phase